lib/std/zig/system/NativeTargetInfo.zig

```diff
@@ -355,37 +355,57 @@ fn detectAbiAndDynamicLinker(
 
-    const env_file = std.fs.openFileAbsoluteZ("/usr/bin/env", .{}) catch |err| switch (err) {
-        error.NoSpaceLeft => unreachable,
-        error.NameTooLong => unreachable,
-        error.PathAlreadyExists => unreachable,
-        error.SharingViolation => unreachable,
-        error.InvalidUtf8 => unreachable,
-        error.BadPathName => unreachable,
-        error.PipeBusy => unreachable,
-        error.FileLocksNotSupported => unreachable,
-        error.WouldBlock => unreachable,
-        error.FileBusy => unreachable, // opened without write permissions
-
-        error.IsDir,
-        error.NotDir,
-        error.InvalidHandle,
-        error.AccessDenied,
-        error.NoDevice,
-        error.FileNotFound,
-        error.FileTooBig,
-        error.Unexpected,
-        => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
+    const elf_file = blk: {
+        // This block looks for a shebang line in /usr/bin/env,
+        // if it finds one, then instead of using /usr/bin/env as the ELF file to examine, it uses the file it references instead,
+        // doing the same logic recursively in case it finds another shebang line.
+
+        // Since /usr/bin/env is hard-coded into the shebang line of many portable scripts, it's a
+        // reasonably reliable path to start with.
+        var file_name: []const u8 = "/usr/bin/env";
+        while (true) {
+            const file = std.fs.openFileAbsolute(file_name, .{}) catch |err| switch (err) {
+                error.NoSpaceLeft => unreachable,
+                error.NameTooLong => unreachable,
+                error.PathAlreadyExists => unreachable,
+                error.SharingViolation => unreachable,
+                error.InvalidUtf8 => unreachable,
+                error.BadPathName => unreachable,
+                error.PipeBusy => unreachable,
+                error.FileLocksNotSupported => unreachable,
+                error.WouldBlock => unreachable,
+                error.FileBusy => unreachable, // opened without write permissions
+
+                error.IsDir,
+                error.NotDir,
+                error.InvalidHandle,
+                error.AccessDenied,
+                error.NoDevice,
+                error.FileNotFound,
+                error.FileTooBig,
+                error.Unexpected,
+                => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
 
-        else => |e| return e,
+                else => |e| return e,
+            };
+
+            // #! (2) + 255 (max length of shebang line since Linux 5.1) + \n (1)
+            var buffer: [258]u8 = undefined;
+            const line = file.reader().readUntilDelimiter(&buffer, '\n') catch break :blk file;
+            if (mem.startsWith(u8, line, "#!")) {
+                if (line[2..].len == 0) return defaultAbiAndDynamicLinker(cpu, os, cross_target);
+                var it = std.mem.tokenize(u8, line[2..], " ");
+                file_name = it.next() orelse line[2..];
```

Comment:
What I meant was doing something like:
```suggestion
                var it = std.mem.tokenize(u8, line[2..], " ");
                file_name = it.next() orelse {
                    file.close();
                    return defaultAbiAndDynamicLinker(cpu, os, cross_target);
                }
```

It would probably be better to use `defer` with a guard like:
```zig
var should_close = true;
defer if (should_close) file.close();
```
And then set `should_close = false` breaking from the block.

Reply:
> What I meant was doing something like:
> 
> It would probably be better to use `defer` with a guard like:
> 
> ```zig
> var should_close = true;
> defer if (should_close) file.close();
> ```
> 
> And then set `should_close = false` breaking from the block.

Done!